Interop/StubGenerator/src/main/kotlin/org/jetbrains/kotlin/native/interop/gen/StubIrMetadataEmitter.kt

```diff
@@ -0,0 +1,333 @@
+/*
+ * Copyright 2010-2019 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license
+ * that can be found in the LICENSE file.
+ */
+package org.jetbrains.kotlin.native.interop.gen
+
+import kotlinx.metadata.*
+import kotlinx.metadata.klib.*
+import org.jetbrains.kotlin.backend.common.serialization.cityHash64
+import org.jetbrains.kotlin.utils.addIfNotNull
+
+class StubIrMetadataEmitter(
+        private val context: StubIrContext,
+        private val builderResult: StubIrBuilderResult,
+        private val moduleName: String
+) {
+    fun emit(): KlibModuleMetadata {
+        val annotations = emptyList<KmAnnotation>()
+        val fragments = emitModuleFragments()
+        return KlibModuleMetadata(moduleName, fragments, annotations)
+    }
+
+    private fun emitModuleFragments(): List<KmModuleFragment> =
+        ModuleMetadataEmitter(context.validPackageName).let {
+            builderResult.stubs.accept(it, null)
+            listOf(it.writeModule())
+        }
+}
+
+private interface StubIrToMetadataTranslationExtensions {
+
+    val FunctionStub.flags: Flags
+        get() = listOfNotNull(
+                Flag.Common.IS_PUBLIC,
+                Flag.Function.IS_EXTERNAL,
+                Flag.HAS_ANNOTATIONS.takeIf { annotations.isNotEmpty() }
+        ).let { flagsOf(*it.toTypedArray()) }
+
+    val Classifier.fqNameSerialized: String
+        get() = fqName.replace('.', '/')
+
+    val PropertyStub.flags: Flags
+        get() = listOfNotNull(
+                Flag.IS_PUBLIC,
+                Flag.Property.IS_DECLARATION,
+                Flag.IS_FINAL,
```

Comment:
This flag better be induced from `this.modality`. Same for `FunctionStub.flags`.